test/functional/session_controller_test.rb

```diff
@@ -0,0 +1,116 @@
+require "test_helper"
+require "gds_api/test_helpers/account_api"
+
+class SessionsControllerTest < ActionController::TestCase
+  include GdsApi::TestHelpers::AccountApi
+  include SessionHelpers
+
+  context "GET sign-in" do
+    setup do
+      stub_account_api_get_sign_in_url(redirect_path: "/bank-holiday")
+    end
+
+    should "redirect the user to the GOV.UK Account service domain" do
+      get :create, params: { redirect_path: "/bank-holiday", state_id: "state123" }
```

Comment:
How does this `:create`, `:callback`, and `:delete` work?  Is it inferring the controller from the name of the class?

Reply:
🤔 so we're inheriting here from `ActionController::TestCase`.
The rails docs for this[ show use of symbols](https://api.rubyonrails.org/classes/ActionController/TestCase.html) that all match the method names of a controller.

My assumption is there's something in `ActionController::TestCase` that matches to a controller file based on the test name... and then uses the symbols to call something like `send(:create)`.

As it's a controller test for function i'd guess this skips the routes alltogether... 

Not really sure let me go take a look around in ActionController::TestCase...

Reply:
The top of the docs do say:
> Rails discourages the use of functional tests in favor of integration tests (use ActionDispatch::IntegrationTest).

But when I was floating Rspec rewrites here the general consensus seemed to be that it could be more confusing having tests breaking from the norm here...

So i'm not sure if these controller tests should exist because... that's what we've done so far in frontend... or if because of their weakness around things like routes we shouldn't be writing any more, even in minitest and we should migrate this over to the integration/request tests..?

Reply:
> So i'm not sure if these controller tests should exist because... that's what we've done so far in frontend...

I think that's a good reason to add them.  Maybe we should migrate frontend over to using the integration tests, as a partial improvement for this repo, but that can come later.

Reply:
Looks like it might get the[ path from here](https://github.com/rails/rails/blob/5aaaa1630ae9a71b3c3ecc4dc46074d678c08d67/actionpack/lib/action_controller/test_case.rb#L499-L501)...
But then where it tries to define the `@controller`... there's a load of `klass` milarky going on in there too.